server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstraction.java

```diff
@@ -178,14 +176,35 @@ public boolean isSystem() {
 
         private final String aliasName;
         private final List<IndexMetadata> referenceIndexMetadatas;
-        private final SetOnce<IndexMetadata> writeIndex = new SetOnce<>();
+        private final IndexMetadata writeIndex;
         private final boolean isHidden;
 
-        public Alias(AliasMetadata aliasMetadata, IndexMetadata indexMetadata) {
+        public Alias(AliasMetadata aliasMetadata, List<IndexMetadata> indices) {
             this.aliasName = aliasMetadata.getAlias();
-            this.referenceIndexMetadatas = new ArrayList<>();
-            this.referenceIndexMetadatas.add(indexMetadata);
+            this.referenceIndexMetadatas = indices;
+
+            List<IndexMetadata> writeIndices = indices.stream()
+                .filter(idxMeta -> Boolean.TRUE.equals(idxMeta.getAliases().get(aliasName).writeIndex()))
+                .collect(Collectors.toList());
+
+            if (writeIndices.isEmpty() && referenceIndexMetadatas.size() == 1
+                && referenceIndexMetadatas.get(0).getAliases().get(aliasName).writeIndex() == null) {
+                writeIndices.add(referenceIndexMetadatas.get(0));
+            }
```

Comment:
I didn't follow this case where the write index is set? It seems to me like it could incorrectly match the scenario when there is a single index for the alias but the index has not been specified as the write index for the alias.

Reply:
I was unaware of this logic too, but if an alias points to a single index then that index is always the write index even if it hasn't been marked as write index. I moved this logic from the `computeAndValidateAliasProperties()` method. I initially also wondered whether this if statement was needed, and didn't move it from the `computeAndValidateAliasProperties()` method, however many tests failed and after moving the if statement the build completed successfully.

Reply:
Ah, good to know. In that case, everything else LGTM.